src/include/gcc.h

```diff
@@ -130,29 +121,29 @@ WT_ATOMIC_CAS_FUNC(size, size_t *vp, size_t old, size_t new)
 static inline bool
 __wt_atomic_cas_ptr(void *vp, void *old, void *new)
 {
-	return (WT_ATOMIC_CAS((void **)vp, old, new));
+	return (WT_ATOMIC_CAS((void **)vp, &old, new));
 }
 
 #define	WT_ATOMIC_FUNC(name, ret, vp_arg, v_arg)			\
 static inline ret							\
 __wt_atomic_add##name(vp_arg, v_arg)					\
 {									\
-	return (__sync_add_and_fetch(vp, v));				\
+	return (__atomic_add_fetch(vp, v, __ATOMIC_SEQ_CST));		\
 }									\
 static inline ret							\
 __wt_atomic_fetch_add##name(vp_arg, v_arg)				\
 {									\
-	return (__sync_fetch_and_add(vp, v));				\
+	return (__atomic_fetch_add(vp, v, __ATOMIC_SEQ_CST));		\
 }									\
-static inline ret							\
+static inline void							\
 __wt_atomic_store##name(vp_arg, v_arg)					\
 {									\
-	return (__sync_lock_test_and_set(vp, v));			\
+	__atomic_store_n(vp, v, __ATOMIC_SEQ_CST);			\
```

Comment:
This is the only change where the actual behaviour may be different.

`__sync_lock_test_and_set` doesn't always use the value passed in. On some architectures, the value just gets set to some implementation defined "on" value. It also uses a weaker "acquire" memory order rather than the sequentially consistent one I'm using here. The analogous function in the `__atomic` series of intrinsics is `__atomic_test_and_set`. However, this doesn't allow us to specify a value that "might" be used like the `__sync` version.

I assumed we probably didn't intend for that subtlety so I've just gone for `__atomic_store_n` with a stronger ordering.

Reply:
I'm kind of confused why we're using `__wt_atomic_store` at all. In the only two places we're using it, we're discarding any return value, which means all it's doing is a store followed by a flush? Do you understand what that's buying us?

Reply:
Likewise, this doesn't make a lot of sense to me.  We assume all over the place that we can do ordinary assignments to variables that are being read and atomically incremented concurrently: I don't see what `__wt_atomic_store` is buying us.

Reply:
I can remove it if you'd like but I don't get why we don't need it.
Don't you need to use an atomic store to avoid a reader thread observing a partially written value? In one of the cases it's a store on a 64 bit value so, to my understanding, on some architectures it might get compiled down to multiple separate stores.

Reply:
WiredTiger requires the architecture not use multiple separate stores for integral or pointer values, there's some discussion here: https://github.com/wiredtiger/wiredtiger/blob/master/src/include/gcc.h#L27-L90

Reply:
fefe1b6

Thanks, that was an interesting read.
Removed them.